File: java/src/main/java/br/com/emendes/problems/NQueens.java
package br.com.emendes.problems;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * O quebra-cabeça n-queens é o problema de colocar n rainhas em um tabuleiro de xadrez nxn
 * de modo que não haja duas rainhas atacando uma à outra.<br><br>
 * <p>
 * Dado um número inteiro n, retorne todas as soluções distintas para o quebra-cabeça das n-rainhas.
 * Você pode retornar a resposta em qualquer ordem.<br><br>
 * <p>
 * Cada solução contém uma configuração de tabuleiro distinta da colocação das n-rainhas,
 * onde 'Q' e '.' ambos indicam uma rainha e um espaço vazio, respectivamente.<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= n <= 9
 */
public class NQueens {

  private List<List<String>> answer;
  private int boardSize;

  public List<List<String>> solveNQueens(int n) {
    this.answer = new ArrayList<>();
    if (n == 1) {
      List<String> board = new ArrayList<>();
      board.add("Q");
      answer.add(board);
      return answer;
    } else if (n == 2 || n == 3) {
      return answer;
    }
    this.boardSize = n;

    putQueenOnLine(generateEmptyLine(n), 0, new ArrayList<>());

    return answer;
  }

  private String[] generateEmptyLine(int n) {
    String[] emptyLine = new String[n];

    Arrays.fill(emptyLine, ".");

    return emptyLine;
  }

  private void putQueenOnLine(String[] line, int currentLine, List<String> currentBoard) {
    if (currentLine == this.boardSize && currentBoard.size() == this.boardSize) {
      this.answer.add(currentBoard);
    }
    String[] lineBelow = generateEmptyLine(this.boardSize);
    validatePositions(line, lineBelow);

    for (int i = 0; i < this.boardSize; i++) {
      if (line[i].equals(".")) {
        // copy line
        String[] lineCopy = Arrays.copyOf(line, this.boardSize);
        // put Q at i
        lineCopy[i] = "Q";
        // invalidate positions i-1, i and i+1 from lineBelow
        String[] lineBelowCopy = Arrays.copyOf(lineBelow, this.boardSize);
        invalidatePosition(i - 1, "l", lineBelowCopy);
        invalidatePosition(i + 1, "r", lineBelowCopy);
        invalidatePosition(i, "b", lineBelowCopy);

        // copy currentBoard
        List<String> currentBoardCopy = new ArrayList<>(currentBoard);
        // save lineCopy at currentBoardCopy
        currentBoardCopy.add(lineToString(lineCopy));
        // call putQueenOnLine.
        putQueenOnLine(lineBelowCopy, currentLine + 1, currentBoardCopy);
      }
    }
  }

  private String lineToString(String[] lineArray) {
    StringBuilder line = new StringBuilder();
    for (int i = 0; i < this.boardSize; i++) {
      if (!lineArray[i].equals("Q")) {
        line.append(".");
      } else {
        line.append("Q");
      }
    }
    return line.toString();
  }

  private void validatePositions(String[] currentLine, String[] lineBelow) {
    for (int i = 0; i < this.boardSize; i++) {
      if (currentLine[i].contains("l")) {
        invalidatePosition(i - 1, "l", lineBelow);
      }
      if (currentLine[i].contains("r")) {
        invalidatePosition(i + 1, "r", lineBelow);
      }
      if (currentLine[i].contains("b")) {
        invalidatePosition(i, "b", lineBelow);
      }
    }
  }

  private void invalidatePosition(int position, String invalidationType, String[] line) {
    if (position > -1 && position < this.boardSize) {
      if (line[position].equals(".")) {
        line[position] = invalidationType;
      } else {
        line[position] += invalidationType;
      }
    }
  }

}

File: javascript/problems/NQueens.js
/**
 * O quebra-cabeça das n rainhas é o problema de colocar n rainhas em um tabuleiro
 * de xadrez nx n de forma que nenhuma das duas rainhas ataque uma à outra.
 *
 * Dado um inteiro n, retorne todas as soluções distintas para o quebra-cabeça n-queens.
 * Você pode retornar a resposta em qualquer ordem.
 *
 * Cada solução contém uma configuração de tabuleiro distinta do posicionamento das n rainhas,
 * onde 'Q' e '.' indicam uma rainha e um espaço vazio, respectivamente.
 *
 * Restrições:
 *
 * 1 <= n <= 9
 *
 * @param {number} n
 * @return {string[][]}
 */
const solveNQueens = function (n) {
  if (n === 2 || n === 3) return [];
  const boards = [];
  const diag1 = createCache(n * 2 - 1);
  const diag2 = createCache(n * 2 - 1);
  const columns = createCache(n);
  console.log(diag1);
  console.log(diag2);
  console.log(columns);
  generateBoards(0, 0, n, diag1, diag2, columns, createEmptyBoard(n), boards);

  return boards;
};

const generateBoards = (
  counter,
  row,
  total,
  diag1,
  diag2,
  columns,
  board,
  boards
) => {
  if (counter === total) {
    boards.push(copyBoard(board));
    return;
  }
  for (let j = 0; j < total; j++) {
    const index = row - j + total - 1;
    if (diag1[index] && diag2[row + j] && columns[j]) {
      diag1[index] = false;
      diag2[row + j] = false;
      columns[j] = false;
      board[row][j] = "Q";
      generateBoards(
        counter + 1,
        row + 1,
        total,
        diag1,
        diag2,
        columns,
        board,
        boards
      );
      diag1[index] = true;
      diag2[row + j] = true;
      columns[j] = true;
      board[row][j] = ".";
    }
  }
};

const copyBoard = (board) => {
  const newBoard = [];
  for (const row of board) {
    newBoard.push(row.join(""));
  }

  return newBoard;
};

const createEmptyBoard = (length) => {
  const board = [];
  for (let i = 0; i < length; i++) {
    const line = new Array(length).fill(".");
    board.push(line);
  }
  return board;
};

const createCache = (length) => {
  return new Array(length).fill(true);
};
